// @ckeditor/ckeditor5-html-support/src/index.js
/**
 * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-licensing-options
 */
/**
 * @module html-support
 */
export { GeneralHtmlSupport } from './generalhtmlsupport.js';
export { DataFilter } from './datafilter.js';
export { DataSchema } from './dataschema.js';
export { HtmlComment } from './htmlcomment.js';
export { FullPage } from './fullpage.js';
export { HtmlPageDataProcessor } from './htmlpagedataprocessor.js';
export { EmptyBlock } from './emptyblock.js';
export { viewToModelObjectConverter as _viewToModelObjectContentHtmlSupportConverter, toObjectWidgetConverter as _toObjectWidgetHtmlSupportConverter, createObjectView as _createObjectHtmlSupportView, viewToAttributeInlineConverter as _viewToAttributeInlineHtmlSupportConverter, emptyInlineModelElementToViewConverter as _emptyInlineModelElementToViewHtmlSupportConverter, attributeToViewInlineConverter as _attributeToInlineHtmlSupportConverter, viewToModelBlockAttributeConverter as _viewToModelBlockAttributeHtmlSupportConverter, modelToViewBlockAttributeConverter as _modelToViewBlockAttributeHtmlSupportConverter } from './converters.js';
export { getDescendantElement as _getHtmlSupportDescendantElement } from './integrations/integrationutils.js';
export { defaultConfig as _HTML_SUPPORT_SCHEMA_DEFINITIONS } from './schemadefinitions.js';
export { updateViewAttributes as _updateHtmlSupportViewAttributes, setViewAttributes as _setHtmlSupportViewAttributes, removeViewAttributes as _removeHtmlSupportViewAttributes, mergeViewElementAttributes as _mergeHtmlSupportViewElementAttributes, modifyGhsAttribute as _modifyHtmlSupportGhsAttribute, toPascalCase as _toHtmlSupportPascalCase, getHtmlAttributeName as _getHtmlSupportAttributeName } from './utils.js';
import './augmentation.js';
